File: src/ISADotNet.Json/Decode.fs
namespace ISADotNet.Json

#if FABLE_COMPILER
open Thoth.Json
#else
open Thoth.Json.Net
open Newtonsoft.Json.Linq
#endif

open ISADotNet
open Fable.Core

module GDecode =

    let isURI (s : string) = 
        true
        //s.StartsWith("http://") || s.StartsWith("https://")

    let uri s json : Result<URI,DecoderError>= 
        match Decode.string s json with
        | Ok s when isURI s -> Ok s
        | Ok s -> Error (DecoderError(s,ErrorReason.FailMessage (sprintf "Expected URI, got %s" s)))
        | Error e -> Error e

    let fromString (decoder : Decoder<'a>) (s : string) : 'a = 
        match Decode.fromString decoder s with
        | Ok a -> a
        | Error e -> failwith (sprintf "Error decoding string: %s" e)
   
    let getFieldNames (json : JsonValue) = 
        match json with
        | :? JObject as json -> 
            json.Properties()
            |> Seq.map (fun x -> x.Name)
        | _ -> Seq.empty

    let hasUnknownFields (knownFields : string list) (json : JsonValue) = 
        getFieldNames json
        |> Seq.exists (fun x -> not (knownFields |> Seq.contains x))

File: ts/ISA/ISA.Json/Decode.ts
import { fromString as fromString_1, string } from "../../fable_modules/Thoth.Json.10.1.0/Decode.fs.js";
import { FSharpResult$2_Ok, FSharpResult$2_Error, FSharpResult$2_$union } from "../../fable_modules/fable-library-ts/Choice.js";
import { ErrorReason_FailMessage, ErrorReason_$union } from "../../fable_modules/Thoth.Json.10.1.0/Types.fs.js";
import { printf, toText } from "../../fable_modules/fable-library-ts/String.js";
import { contains, exists } from "../../fable_modules/fable-library-ts/Seq.js";
import { stringHash } from "../../fable_modules/fable-library-ts/Util.js";
import { int32 } from "../../fable_modules/fable-library-ts/Int32.js";
import { FSharpList } from "../../fable_modules/fable-library-ts/List.js";

export function isURI(s: string): boolean {
    return true;
}

export function uri<$a>(s: string, json: $a): FSharpResult$2_$union<string, [string, ErrorReason_$union]> {
    const matchValue: FSharpResult$2_$union<string, [string, ErrorReason_$union]> = string(s, json);
    if (matchValue.tag === /* Error */ 1) {
        return FSharpResult$2_Error<string, [string, ErrorReason_$union]>(matchValue.fields[0]);
    }
    else if (isURI(matchValue.fields[0])) {
        return FSharpResult$2_Ok<string, [string, ErrorReason_$union]>(matchValue.fields[0]);
    }
    else {
        const s_3: string = matchValue.fields[0];
        return FSharpResult$2_Error<string, [string, ErrorReason_$union]>([s_3, ErrorReason_FailMessage(toText(printf("Expected URI, got %s"))(s_3))] as [string, ErrorReason_$union]);
    }
}

export function fromString<a>(decoder: ((arg0: string, arg1: any) => FSharpResult$2_$union<a, [string, ErrorReason_$union]>), s: string): a {
    let arg: string;
    const matchValue: FSharpResult$2_$union<a, string> = fromString_1<a>(decoder, s);
    if (matchValue.tag === /* Error */ 1) {
        throw new Error((arg = matchValue.fields[0], toText(printf("Error decoding string: %s"))(arg)));
    }
    else {
        return matchValue.fields[0];
    }
}

export function getFieldNames<$a>(json: any): $a {
    return Object.getOwnPropertyNames(json);
}

export function hasUnknownFields(knownFields: FSharpList<string>, json: any): boolean {
    return exists<string>((x: string): boolean => !contains<string>(x, knownFields, {
        Equals: (x_1: string, y: string): boolean => (x_1 === y),
        GetHashCode: stringHash,
    }), getFieldNames<Iterable<string>>(json));
}

